### PR TITLE
refactor(recipes): delta-encode the recipe snapshot fixture (#638)

### LINKED ISSUE #638: Recipe-config snapshot: a schema-default change reports 149 identical failures, and an added field is indistinguishable from a changed value
The recipe-config snapshot from #621 / #637 works and catches the exposure it was built for — changing one schema default turns 149 named recipe tests red, where `main` turns nothing red. This issue is about its **shape**, not its correctness, and it is a refinement rather than a defect. Filed at merge time so it is not rediscovered in three releases.

## The cost that is permanent

The fixture stores a fully resolved config per recipe: 162 recipes x ~296 flattened fields, 49,586 lines, 1.55 MB. Almost all of it is the same defaults repeated 162 times.

Two consequences, both measured on #637's branch:

1. **Adding one `TrainingConfig` field reddens all 162 recipe tests** (measured: 164 failed, 1 passed, with a probe field added). This project adds schema fields most releases, so `162 red -> regenerate -> green` becomes a routine reflex.
2. **Once it is a reflex, an additive change is procedurally indistinguishable from a value change.** "I added a harmless field" and "I changed `epochs` 3 -> 1 and silently retuned 36 recipes" produce the same wall of red and the same fix. The failure message *does* distinguish them — an added field reads `('<missing>', False)` while a moved value reads `(0.1, 0.2)` — so a careful reader can tell, but nothing structural separates the two classes, and a guard that depends on care being applied every time is a guard with a half-life.

Churn per change, for scale: adding a recipe is `+306` lines; adding a `TrainingConfig` field is `+324 / -162`. Both are reviewable. It is the *total* that is 49,586 lines, and it regenerates wholesale on any schema change.

## Proposed shape

Store the resolved defaults **once**, and per recipe store only the fields that differ from them:

- roughly 15x smaller, by inspection of how much of the fixture is repeated defaults
- a schema-default change becomes **one** failure ("the shared defaults moved") rather than 149 identical ones
- an added field with a default touches the defaults block only, so the additive case stops reddening every recipe and becomes visibly different from a value change
- a recipe whose value genuinely moves still fails by name, which is the property that must not be lost

## What must not regress

Every one of these is verified working today and each should be re-verified against any replacement, by mutation rather than by reading:

- [ ] changing a schema default fails, naming the affected recipes
- [ ] changing a recipe value that is not a default fails by name
- [ ] deleting a line that duplicates a default still changes nothing (exposure 1 is deliberately unpinned — see #621)
- [ ] an empty fixture fails loudly; a deleted fixture is a collection error
- [ ] a recipe added without regenerating fails with the recipe name
- [ ] the failure message still names the recipe, the dotted field path, both values, and the regeneration script
- [ ] regeneration remains byte-exact (`git status --porcelain` empty after running the script)

## Acceptance

- [ ] fixture is materially smaller and the number is stated rather than estimated
- [ ] a schema-default change produces one failure, not one per recipe
- [ ] the additive case and the value-change case are distinguishable **structurally**, not only by reading the diff text
- [ ] every box in the section above still passes, each demonstrated by a named mutation

Not urgent, and explicitly **not** a blocker on #637, which is correct as it stands. Whoever takes this should read #621 first for why the snapshot pins resolved behaviour rather than YAML source text — that constraint is the whole reason the instrument looks like this.

### PR DESCRIPTION
**Pull request by SID-6921**
## Summary

Closes #638.

**Stacked on #637**, which hasn't merged yet — so this diff currently includes #637's commits too. The only commit that's actually #638's own content is the last one, `58333e4` ("delta-encode the recipe snapshot fixture"); everything before it is #637. Will rebase onto `main` and force-push once #637 lands, at which point this PR's diff will shrink to just the #638 change. Flagging this up front rather than letting it look like a much bigger PR than it is.

#621/#637's fixture stored each recipe's full resolved config: 162 copies of the same ~296 schema defaults, 49,586 lines / 1,571,967 bytes. A schema-default change reddened every recipe relying on it (149, in the `dpo_beta` case) identically to how an added field reddened all 162 — the same wall of red for two different classes of change.

## What's here

- Shared defaults stored once, as a `baseline` (a minimal resolved config: placeholder base/task/data, everything else at schema default).
- Per recipe, only the flattened fields that differ from that baseline.
- `TestBaselineMatchesTheLiveSchema` — the one place a schema-default change or an added field shows up now, not once per recipe.
- Per-recipe tests compare each recipe's delta against the *live* baseline, not the committed one, so a schema-default change alone doesn't move them for recipes that don't touch the changed field.
- `generate_recipe_snapshot.py` and the test import the same `flatten()` / `_delta()` / `build_baseline()` — no duplication between what builds the fixture and what verifies it, continuing the fix from the #637 review round.

Measured: fixture goes from 1,571,967 bytes / 49,586 lines to 68,600 bytes / 2,040 lines — 22.9x smaller by byte count, 24.3x fewer lines.

## One nuance, not smoothed over

A recipe that redundantly pins a value equal to the *old* default starts appearing in its delta once the default moves out from under it — because what used to be a no-op pin just started doing something. Verified on the real catalog: the `dpo_beta` 0.1→0.2 mutation produces **14** failures (the baseline, plus 13 named recipes with that exact pin), not literally one — down from 149, and each of the 14 is individually meaningful rather than 149 duplicates of the same signal.

## Verification — full checklist from the issue, each by mutation against the real catalog

- schema default change (`dpo_beta` 0.1→0.2): **14** failures, not 149
- added schema field: **1** failure (the baseline), not 164
- a recipe's own value changing (`lora.r` 16→8): 1 named failure
- exposure 1 (deleting a redundant default-duplicating line): no failures, unchanged from #621
- recipe added without regenerating: 2 named failures (completeness check + per-recipe), not a `KeyError`
- empty fixture (`{}`): collection error, not a per-test failure
- deleted fixture: collection error
- regeneration: byte-exact (`git status --porcelain` empty after running `scripts/generate_recipe_snapshot.py`, including on Windows — carries forward #637's `newline="\n"` fix)

166 tests passing, ruff and mypy clean.

### CODE REVIEW

File: scripts/generate_recipe_snapshot.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    sys.path.insert(0, str(_REPO_ROOT / "src"))
+    from soup_cli.config.loader import load_config_from_string
+
+    return flatten(load_config_from_string(_BASELINE_YAML).model_dump(mode="json"))
+
+
+def build_snapshot() -> dict[str, dict]:
+    sys.path.insert(0, str(_REPO_ROOT / "src"))
+    from soup_cli.config.loader import load_config_from_string
+    from soup_cli.recipes.catalog import RECIPES
```
[the remark this answers is not in the document]
Comment (SID-6921) [reply scripts/generate_recipe_snapshot.py:90]: Fixed — added _ensure_src_on_path(), checks sys.path membership before inserting. build_snapshot() calling build_baseline() internally was exactly the path that could double it up. Verified: sys.path has exactly one entry for the src dir after calling both functions repeatedly.

File: tests/test_issue621_recipe_snapshot.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+from __future__ import annotations
+
+import functools
+import json
+from pathlib import Path
+
+import pytest
+
+from scripts.generate_recipe_snapshot import _delta, build_baseline, flatten
```
[the remark this answers is not in the document]
Comment (SID-6921) [reply tests/test_issue621_recipe_snapshot.py]: Renamed to delta (public) in the generator; the test now imports the public name too. Agreed it was signaling private while being used cross-module.

File: tests/test_issue621_recipe_snapshot.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+_REGENERATE_HINT = "regenerate with scripts/generate_recipe_snapshot.py"
+_MISSING = object()
+
+
+@functools.lru_cache(maxsize=1)
+def _load_fixture() -> dict:
+    data = json.loads(_FIXTURE_PATH.read_text(encoding="utf-8"))
+    assert isinstance(data, dict) and "baseline" in data and "recipes" in data, (
+        f"fixture must be a dict with 'baseline' and 'recipes' keys; {_REGENERATE_HINT}"
+    )
```
[the remark this answers is not in the document]
Comment (SID-6921) [reply tests/test_issue621_recipe_snapshot.py]: Changed to raise ValueError instead of assert. Verified: ran python -O against a deliberately malformed ({}) fixture — the old assert would have been stripped and let it through silently; the raise still fires with the same message.

File: tests/test_issue621_recipe_snapshot.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    out: dict[str, tuple[object, object]] = {}
+    for path in paths:
+        e = expected.get(path, _MISSING)
+        a = actual.get(path, _MISSING)
+        if e != a:
+            out[path] = (
+                "<missing>" if e is _MISSING else e,
+                "<missing>" if a is _MISSING else a,
+            )
+    return out
```
[the remark this answers is not in the document]
Comment (SID-6921) [reply tests/test_issue621_recipe_snapshot.py]: Changed to sorted(set(...)) before building the diff dict. Verified against an out-of-insertion-order dict that the reported field order is now alphabetical rather than set-iteration-order dependent.

**Review by MakazhanAlpamys — CHANGES_REQUESTED** [review]
Reviewed at `e52e12f`. **Requesting changes**, but the refactor itself is sound: all seven properties #638 said must not regress were re-verified by mutation and all seven hold, and the size reduction is real and larger than claimed.

## Mechanical first: it does not merge

```
$ git merge-tree --write-tree origin/main pr640
CONFLICT (add/add): scripts/generate_recipe_snapshot.py
CONFLICT (add/add): tests/fixtures/recipe_config_snapshots.json
CONFLICT (add/add): tests/test_issue621_recipe_snapshot.py
```

The branch carries #637's unsquashed commits (`07be732`, `be911f4`) while #637 landed on `main` as the squash `e83b992`. **The good news is that the rebase is purely mechanical**: `be911f4:<file>` and `e83b992:<file>` are the same blob SHA for all three files, so the fixture you refactored is byte-identical to the merged one. Rebase onto `e83b992` and take your side.

## The measured numbers, and yours is understated

| | main | #640 | ratio |
|---|---|---|---|
| committed blob | 1,571,967 B | **66,560 B** | **23.6x** |
| lines | 49,586 | **2,040** | 24.3x |

Your fragment says `68,600 bytes` and `22.9x`. The difference is exactly 2,040 — one byte per line — so it was measured on a CRLF working copy before `.gitattributes eol=lf` applied. Worth correcting because it understates your own result.

## The finding: a new false positive on redundantly-pinned defaults

I ran the same schema-default mutation on both branches:

| default mutated | main | #640 |
|---|---|---|
| `warmup_ratio` 0.03 -> 0.07 (0 recipes pin it) | 162 failed | **1** — baseline only |
| `dpo_beta` 0.1 -> 0.2 | 149 failed | **14** — baseline + the 13 that pinned it |
| **`epochs` 3 -> 1** | **0 failed** | **159 failed** |
| new field added | 162 failed | **1** |

The first, second and fourth rows are #638 working exactly as asked. The third is a new class: **not one recipe's resolved config changes** — a recipe declaring `epochs: 3` still resolves to 3 — yet 158 named recipe tests go red because the pin fell out of the shared baseline (`('<missing>', 3)`). Main is correctly silent there. That is the wall of red #638 set out to remove, reappearing on a change that alters nothing.

`tests/test_issue621_recipe_snapshot.py:28-33` acknowledges the mechanism and calls it "a real, if small, class of failure". **158 of 162 recipes is not small**, and the fragment reports only the flattering `dpo_beta` case (14) while omitting this one. I would rather the fragment stated the worst case than the best; it is the number a maintainer needs when a schema PR turns the suite red.

Related and deliberate rather than a regression: "changing a schema default fails, **naming the affected recipes**" is now traded away for defaults nobody pins — `warmup_ratio` gives one failure naming the *field* and zero recipes. #638's acceptance criterion 2 asked for exactly that, so I read the trade as intended; the two bullets in my own issue did conflict, and you resolved it the right way. Say so in the fragment rather than leaving it implicit.

## The "structurally distinguishable" claim is not supported

Adding a field gives 1 failure in `TestBaselineMatchesTheLiveSchema` showing `('<missing>', False)`. Changing an unpinned default gives **1 failure in the same test** showing `(0.03, 0.07)`. Same test, same count — separated only by reading the message, which is precisely the discriminator #638 said must not be the one. Your fragment's sentence "an added field is now structurally distinguishable from a changed value rather than only distinguishable by reading the diff text" needs to go or be narrowed.

What you *did* make structural is real and worth claiming instead: **"a recipe's own value moved" (a named recipe test) versus "the shared defaults moved" (the baseline test)** are now different tests. That is a genuine gain and it is the honest version of the sentence.

## Everything else checks out

All seven properties, by mutation: recipe value that is not a default fails by name; deleting a line duplicating a default still kills nothing (exposure 1 correctly unpinned); `{}` fixture, empty file and deleted file each fail loudly at collection with a message naming the regeneration script; a recipe added without regenerating fails twice, both naming it; the message still carries recipe, dotted path, both values and the script; regeneration byte-exact with `git status --porcelain` empty. `newline="\n"` and `.gitattributes` both survive. 162 deltas against a 162-recipe catalog. ruff clean, `test_cli_startup_is_light` 34 passed, 166 tests in the file.

**57 deleted test lines, all read**: docstring rewrite, `_flatten` -> public `flatten` moved into the script, `_live_snapshots()` removed, `_resolve` -> `_resolve_flat`, the class rename, and the two vacuity controls rewritten from nested to flat dicts. Nothing silently dropped — 165 tests to 166.

One deletion I want to name without blocking on it. #637's docstring recorded *why* the test calls `build_snapshot()` rather than re-implementing resolution: two independent copies of `model_dump(mode="json")` is the shape this repo has been bitten by in #372 and #392. #640 deletes that paragraph and re-implements per-recipe resolution at `:84-87`. I probed for a live gap three ways rather than asserting one — `mode="json"` -> `mode="python"` and `exclude_none=True` in the script alone are both **equivalent** (regenerated fixture byte-identical, 166 passed), while injecting an extra delta key is caught loudly (162 failed). So the divergence mode is self-announcing under this design and I found no hole. What was lost is the recorded reasoning, and this project keeps re-learning that lesson because the reasoning gets deleted, not because the code was wrong.

**To unblock:** rebase onto `e83b992`, correct the byte count and the ratio, replace the "structurally distinguishable" sentence with the recipe-versus-baseline claim that is actually true, and state the `epochs`-shaped worst case rather than the `dpo_beta` best case.

**Comment by SID-6921** [comment]
Rebased and corrected in 3e05aa1.

Rebase: cherry-picked the three commits that are actually #638's own work onto e83b992 rather than replaying the full branch — the earlier commits duplicate what #637 already landed as one squash commit, so a linear rebase conflicted on content that's already there. Clean apply, confirms the blobs matched as you said. PR now shows mergeable.

Numbers: fixed to 66,560 bytes / 23.6x — the 68,600 figure was measured on a CRLF working copy before .gitattributes normalized it.

"Structurally distinguishable": pulled the claim about added-field vs changed-value, since your testing showed it doesn't hold (same test, same failure count, only the message text differs). Replaced with the claim that's actually true: "recipe's own value moved" and "shared defaults moved" are different tests now, which they weren't before.

epochs case: reproduced it myself — 159 failures (baseline + all 158 recipes that declare it explicitly), not a small edge case. Both that number and the dpo_beta one are in the fragment and the module docstring now, with the trade stated as deliberate rather than implied to be rare.

**Review by MakazhanAlpamys — CHANGES_REQUESTED** [review]
**All three findings closed, and the encoding is provably lossless — I verified that rather than accepting it. One factual line in the fragment stands between this and merge.**

## The question I actually cared about: is 149 -> 14 a loss of coverage?

No, and it is demonstrable rather than arguable. I reconstructed every recipe from `baseline + delta` using the PR's own `flatten`:

```
recipes: 162 | reconstruction mismatches: 0
key union new: 296 | old flattened: 296 | equal: True
```

Zero keys lost, zero gained, and every one of the 162 recipes rebuilds to exactly what `main`'s fixture holds. The file stores identical information, deduplicated. So the `dpo_beta` mutation going from 149 failures to 14 is a change in *reporting shape*, not in what is detected — and the two failing sets being disjoint (main's 149 are the recipes that do not pin it, yours are the 13 that do) is the cleanest possible evidence of that.

One thing I should say plainly: **my first attempt at this check reported 162 mismatches**, because I compared the flat deltas against the old fixture's *nested* structure. That was my error, not a finding, and I mention it because a reviewer who stopped there would have blocked a correct PR on their own bug.

## Findings A and B

**A — the rebase is lossless.** `merge-base` is `2285b65`, an ancestor of `main`, and `merge-tree` produces no conflict. The blob SHAs for `scripts/generate_recipe_snapshot.py` and `tests/fixtures/recipe_config_snapshots.json` are **identical** across your old and new heads, so the rebase changed exactly the two files I asked about and left the other two bit-for-bit. Cherry-picking the three commits that are #638's own work rather than replaying the branch was the right call.

**B — the numbers are corrected and I re-measured them.** `1,571,967 -> 66,560` bytes is **23.6x**, `49,586 -> 2,040` lines is 24.3x. The fragment now says exactly that. The CRLF trap did not fire.

**And the honesty upgrade is the best part of this round.** The fragment now states the `epochs` worst case — a near-universally-pinned field producing a wall of named failures — instead of showing only the favourable `dpo_beta` example. That is a trade-off you were not obliged to volunteer, stated at its real size, and it is what makes the rest of the fragment trustworthy.

## The one thing to fix

> `epochs`, 3 -> 1, **which 158 of 162 recipes declare explicitly**

Measured against the live catalogue:

```
total recipes:                162
declare epochs explicitly:    162        <- all of them
distribution:                 {1: 32, 2: 4, 3: 126}
```

The number 158 is right; the reason is not. It is not "158 declare it" — it is **126 that declare the old default plus 32 that declare the new one**, both of which move relative to the baseline. The 4 recipes declaring `2` stay green precisely because their value is neither default.

That distinction is the whole predictive value of the sentence. The real rule is *"a recipe declaring a value equal to the old **or** the new default"*, and it lets a maintainer size the blast radius of a future default change before making it; "158 declare it" does not, and is wrong in a way that would mislead exactly the person who most needs it. It ships verbatim into `CHANGELOG.md`, so it is worth the one line.

**If you would rather not spin another round for one clause, say so and I will edit the fragment myself before merging** — it is your prose and I would rather ask than rewrite it silently.

## Verification

`tests/test_issue621_recipe_snapshot.py` — 166 passed. Regeneration byte-exact: `git status --porcelain` empty afterwards, 0 CRLF. `ruff` clean. **56 deleted test lines, all read individually** — docstring rewrite, `_flatten` moved into the script as the public `flatten`, `_diff` rewritten flat, a class rename, and the vacuity controls' dicts flattened. No assertion removed; test count 165 -> 166. The fixture shrinking by 51,618 lines is the feature.

All seven of #638's must-not-regress properties re-verified: schema-default change fails; a recipe's own value failing by name; deleting a redundant default-duplicating line still kills nothing (exposure 1 correctly left unpinned); malformed/empty/deleted fixture each raise at collection naming the script; a recipe added without regenerating fails twice by name; the message carries recipe, dotted path, both values and the script; regeneration byte-exact.

## Three notes for later, none blocking

1. **The `_MISSING` sentinel is unexercised and its branch is reachable.** Replacing `baseline.get(path, _MISSING)` with `baseline.get(path)` survives all 166 tests. It is not merely equivalent: a recipe filling `training.eval_gate` with an unset Optional sub-field yields a path absent from the baseline, which the mutated form silently drops. No recipe does that today — so the sentinel is correct, forward-looking, and untested. Five synthetic lines in the `TestTheSnapshotIsNotVacuous` style would pin it.
2. **The guard now has a single point of detection.** With `TestBaselineMatchesTheLiveSchema` neutered, a `warmup_ratio` default change is *completely* invisible; on main the same change reddened 162 tests, so skipping any one left 161 red. That is #638's criterion 2 working as specified, not a defect — but one docstring sentence saying that test must never be skipped or `xfail`ed is cheap insurance.
3. **No independent oracle for the shared helpers.** The test imports `delta` and `flatten` from the script, so breaking `delta()` *and regenerating* leaves the suite green with a 1.9 MB fixture. Loud in review, inherent to regenerate-from-the-code-under-test, and killable cheaply: assert that no committed delta entry restates a baseline value. That holds today — 0 such entries.

## Housekeeping that is mine

`pytorch-smoke` became a required check after your run, so this shows `BLOCKED` at 13/13 green — the job did not exist at your base. I will update the branch when the fragment line is settled; nothing for you to do about it.

**Comment by SID-6921** [comment]
Fixed in 06cc2be. Measured it myself before touching anything: {1: 32, 2: 4, 3: 126} across all 162 recipes — every recipe declares epochs, none omit it. The 158 is 126 pinning the old default plus 32 already pinning the new one, both now differing from a baseline that used to match one of them; the 4 pinning 2 stay green since neither default was ever their value. Fragment and docstring both corrected to state the actual reason, not just the right number for the wrong one.

Appreciate the offer to fix it yourself, but this one was on me to get right and quick enough to just do.

**Comment by MakazhanAlpamys** [comment]
Merged as `36da5af`, credited in CONTRIBUTORS at `33d6af9`. **#638 closed.**

The correction landed exactly right, and the part I want to name is that **you re-measured the distribution yourself instead of taking my number on trust** — `{1: 32, 2: 4, 3: 126}` across all 162, matching what I had measured independently. A contributor who verifies a maintainer's correction before applying it is doing the thing that catches a maintainer's mistake, and I have made two this week.

The fragment now gives the rule that predicts blast radius rather than a number without a mechanism: 158 is the 126 pinning the old default plus the 32 already pinning the new one, and the 4 pinning `2` stay green because neither default was ever their value.

Verified before merging:

- **Losslessness measured, not argued.** Reconstructing every recipe as `baseline + delta` reproduces the old fixture with **zero mismatches across all 162**, key union 296 either way. So 149 → 1 is a reporting-shape change; nothing stopped being detected.
- 166 tests pass; **regeneration byte-exact** — `git status --porcelain` empty afterwards, and the documented Windows CRLF trap did not fire.
- 14/14 CI including `pytorch-smoke`.
- 56 removed lines outside the fixture, each read: docstring rewrite, `_flatten` moved into the script as public `flatten`, `_diff` rewritten flat, a class rename, vacuity dicts flattened. **No assertion removed**, 165 → 166 tests.
- The property that must **not** fire still does not: deleting a recipe line that merely duplicates a default kills nothing, because #621 excluded that direction.

Volunteering the `epochs` worst case — a wall of 158 named failures on a change that alters no recipe's behaviour — is what makes the rest of the fragment trustworthy. Nobody asked for it.

Three things recorded in the merge commit as known rather than hidden, all fair game later: the `_MISSING` sentinel's branch is reachable but unexercised (replacing it with `.get(path)` survives all 166); the guard now has a single point of detection, so `TestBaselineMatchesTheLiveSchema` must never be skipped or `xfail`ed; and since the test imports `delta` and `flatten` from the script, breaking `delta()` *and* regenerating leaves the suite green with a 1.9 MB fixture — loud in review, inherent to the pattern, and killable by asserting no committed delta restates a baseline value (0 such entries today).

#639 is under review now at `536eed9`.
